polaris/experimental/_dataset_v2.py

```diff
@@ -0,0 +1,249 @@
+import json
+import re
+import uuid
+from pathlib import Path
+from typing import Any, ClassVar, Literal
+
+import fsspec
+import numpy as np
+import zarr
+from loguru import logger
+from pydantic import PrivateAttr, computed_field, model_validator
+
+from polaris.dataset._adapters import Adapter
+from polaris.dataset._base import _CACHE_SUBDIR, BaseDataset
+from polaris.dataset.zarr._manifest import calculate_file_md5, generate_zarr_manifest
+from polaris.utils.constants import DEFAULT_CACHE_DIR
+from polaris.utils.errors import InvalidDatasetError
+from polaris.utils.types import AccessType, HubOwner, ZarrConflictResolution
+
+_INDEX_ARRAY_KEY = "__index__"
+
+
+class DatasetV2(BaseDataset):
+    """Second version of a Polaris Dataset.
+
+    This version gets rid of the DataFrame and stores all data in a Zarr archive.
+
+    V1 stored all datapoints in a Pandas DataFrame. Because a DataFrame is always loaded to memory,
+    this was a bottleneck when the number of data points grew large. Even with the pointer columns, you still
+    need to load all pointers into memory. V2 therefore switches to a Zarr-only format.
+
+    Info: This feature is still experimental
+        The DatasetV2 is in active development and will likely undergo breaking changes before release.
+
+    Attributes:
+        zarr_root_path: The path to the Zarr archive. Different from V1, this is now required.
+
+    For additional meta-data attributes, see the [`BaseDataset`][polaris._dataset.BaseDataset] class.
+
+    Raises:
+        InvalidDatasetError: If the dataset does not conform to the Pydantic data-model specification.
+    """
+
+    version: ClassVar[Literal[2]] = 2
+    _zarr_manifest_path: str | None = PrivateAttr(None)
+    _zarr_manifest_md5sum: str | None = PrivateAttr(None)
+
+    # Redefine this to make it a required field
+    zarr_root_path: str
+
+    @model_validator(mode="after")
+    @classmethod
+    def _validate_v2_dataset_model(cls, m: "DatasetV2"):
```

Comment:
**Suggestion**
Model validator with `after` mode can be instance methods in Pydantic:
```suggestion
    @model_validator(mode="after")
    def _validate_v2_dataset_model(self) -> Self:
```

Reply:
You can with [typing-extensions](https://pypi.org/project/typing-extensions/).